File: lib/api/douyin_api_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../model/models.dart';

class DouyinApiClient {
  static const String _douyinApi = 'https://api.bugpk.com/api/douyin';
  static const String _xhsApi = 'https://api.bugpk.com/api/xhs';

  final http.Client _client;

  DouyinApiClient({http.Client? client}) : _client = client ?? http.Client();

  /// 解析抖音视频链接
  Future<ApiResponse> parseVideo(String shareUrl) async {
    try {
      final response = await _client.post(
        Uri.parse(_douyinApi),
        body: {'url': shareUrl},
        headers: {
          'User-Agent': 'DouyinDownloader/2.0 (Flutter)',
          'Accept': 'application/json',
        },
      );

      if (response.statusCode != 200) {
        throw Exception('HTTP ${response.statusCode}: ${response.body}');
      }

      final json = jsonDecode(response.body) as Map<String, dynamic>;
      final apiResponse = ApiResponse.fromJson(json);

      if (apiResponse.code != 200) {
        throw Exception(apiResponse.msg);
      }

      return apiResponse;
    } catch (e) {
      rethrow;
    }
  }

  /// 解析小红书链接
  Future<XhsApiResponse> parseXiaohongshu(String shareUrl) async {
    try {
      final uri = Uri.parse(_xhsApi).replace(queryParameters: {'url': shareUrl});
      final response = await _client.get(
        uri,
        headers: {
          'User-Agent': 'DouyinDownloader/2.0 (Flutter)',
          'Accept': 'application/json',
        },
      );

      if (response.statusCode != 200) {
        throw Exception('HTTP ${response.statusCode}: ${response.body}');
      }

      final json = jsonDecode(response.body) as Map<String, dynamic>;
      final xhsResponse = XhsApiResponse.fromJson(json);

      if (xhsResponse.code != 200) {
        throw Exception(xhsResponse.msg);
      }

      return xhsResponse;
    } catch (e) {
      rethrow;
    }
  }

  /// 解小红书短链接
  static Future<String> resolveXhsShortLink(String shortUrl) async {
    try {
      final client = http.Client();
      try {
        final request = http.Request('HEAD', Uri.parse(shortUrl));
        request.headers.addAll({
          'User-Agent': 'DouyinDownloader/2.0 (Flutter)',
        });
        final streamedResponse = await client.send(request);
        final location = streamedResponse.headers['location'] ?? '';
        streamedResponse.discard();
        if (location.isNotEmpty) {
          return location.startsWith('http')
              ? location
              : 'https://www.xiaohongshu.com$location';
        }
      } finally {
        client.close();
      }
    } catch (_) {}
    return shortUrl;
  }

  void dispose() {
    _client.close();
  }
}

File: app/src/main/java/com/douyin/downloaderqh/api/DouyinApiClient.kt
package com.douyin.downloaderqh.api

import com.douyin.downloaderqh.model.ApiResponse
import com.douyin.downloaderqh.model.XhsApiResponse
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import okhttp3.*
import java.io.IOException

class DouyinApiClient {

    companion object {
        private const val DOUYIN_API = "https://api.bugpk.com/api/douyin"
        private const val XHS_API = "https://api.bugpk.com/api/xhs"

        val json = Json {
            ignoreUnknownKeys = true
            coerceInputValues = true
        }
    }

    private val client = OkHttpClient.Builder()
        .followRedirects(true)
        .connectTimeout(30, java.util.concurrent.TimeUnit.SECONDS)
        .readTimeout(60, java.util.concurrent.TimeUnit.SECONDS)
        .writeTimeout(60, java.util.concurrent.TimeUnit.SECONDS)
        .build()

    suspend fun parseVideo(shareUrl: String): Result<ApiResponse> = withContext(Dispatchers.IO) {
        try {
            val requestBody = FormBody.Builder()
                .add("url", shareUrl)
                .build()

            val request = Request.Builder()
                .url(DOUYIN_API)
                .post(requestBody)
                .addHeader("User-Agent", "DouyinDownloader/2.0 (Android)")
                .addHeader("Accept", "application/json")
                .build()

            val response = client.newCall(request).execute()
            val bodyString = response.body?.string() ?: throw IOException("响应体为空")

            if (!response.isSuccessful) {
                return@withContext Result.failure(IOException("HTTP ${response.code}: $bodyString"))
            }

            val apiResponse = json.decodeFromString<ApiResponse>(bodyString)

            if (apiResponse.code == 200) Result.success(apiResponse)
            else Result.failure(IOException(apiResponse.msg))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun parseXiaohongshu(shareUrl: String): Result<XhsApiResponse> = withContext(Dispatchers.IO) {
        try {
            val httpUrl = HttpUrl.Builder()
                .scheme("https")
                .host("api.bugpk.com")
                .addPathSegment("api")
                .addPathSegment("xhs")
                .addQueryParameter("url", shareUrl)
                .build()

            val request = Request.Builder()
                .url(httpUrl)
                .get()
                .addHeader("User-Agent", "DouyinDownloader/2.0 (Android)")
                .addHeader("Accept", "application/json")
                .build()

            val response = client.newCall(request).execute()
            val bodyString = response.body?.string() ?: throw IOException("响应体为空")

            if (!response.isSuccessful) {
                return@withContext Result.failure(IOException("HTTP ${response.code}: $bodyString"))
            }

            val xhsResponse = json.decodeFromString<XhsApiResponse>(bodyString)

            if (xhsResponse.code == 200) Result.success(xhsResponse)
            else Result.failure(IOException(xhsResponse.msg))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}
